File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/items/media/files/properties/playstats/playedfile/specs/GivenAConnectionThatReturnsA404ErrorCode/WhenSendingPlayedToServer.java
package com.lasthopesoftware.bluewater.client.library.items.media.files.properties.playstats.playedfile.specs.GivenAConnectionThatReturnsA404ErrorCode;

import com.lasthopesoftware.bluewater.client.connection.specs.FakeConnectionProvider;
import com.lasthopesoftware.bluewater.client.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.library.items.media.files.properties.playstats.playedfile.PlayedFilePlayStatsUpdater;
import com.lasthopesoftware.bluewater.shared.exceptions.HttpResponseException;
import com.lasthopesoftware.bluewater.shared.promises.extensions.specs.FuturePromise;
import org.junit.BeforeClass;
import org.junit.Test;

import java.util.concurrent.ExecutionException;

import static org.assertj.core.api.Assertions.assertThat;

public class WhenSendingPlayedToServer {

	private static HttpResponseException httpResponseException;

	@BeforeClass
	public static void before() throws InterruptedException {
		final FakeConnectionProvider connectionProvider = new FakeConnectionProvider();

		final PlayedFilePlayStatsUpdater updater = new PlayedFilePlayStatsUpdater(connectionProvider);

		try {
			new FuturePromise<>(updater.promisePlaystatsUpdate(new ServiceFile(15))).get();
		} catch (ExecutionException e) {
			if (e.getCause() instanceof HttpResponseException)
				httpResponseException = (HttpResponseException)e.getCause();
		}
	}

	@Test
	public void thenAnHttpResponseExceptionIsThrownWithTheResponseCode() {
		assertThat(httpResponseException.getResponseCode()).isEqualTo(404);
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/browsing/files/properties/playstats/playedfile/GivenAConnectionThatReturnsARandomErrorCode/WhenSendingPlayedToServer.kt
package com.lasthopesoftware.bluewater.client.browsing.files.properties.playstats.playedfile.GivenAConnectionThatReturnsARandomErrorCode

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.files.properties.playstats.playedfile.PlayedFilePlayStatsUpdater
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.connection.FakeConnectionProvider
import com.lasthopesoftware.bluewater.client.connection.FakeConnectionResponseTuple
import com.lasthopesoftware.bluewater.shared.exceptions.HttpResponseException
import com.lasthopesoftware.bluewater.shared.promises.extensions.ProgressingPromise
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import java.util.*
import java.util.concurrent.ExecutionException

private const val libraryId = 72

class WhenSendingPlayedToServer {

	private val expectedResponseCode by lazy {
		val random = Random()
		var responseCode: Int
		do {
			responseCode = random.nextInt()
		} while (responseCode < 0 || responseCode in 200..299)
		responseCode
	}

	private val updater by lazy {
		PlayedFilePlayStatsUpdater(mockk {
			val connectionProvider = FakeConnectionProvider()
			connectionProvider.mapResponse({
				FakeConnectionResponseTuple(
					expectedResponseCode, ByteArray(0)
				)
			}, "File/Played", "File=15", "FileType=Key")

			every { promiseLibraryConnection(LibraryId(libraryId)) } returns ProgressingPromise(connectionProvider)
		})
	}
	private var httpResponseException: HttpResponseException? = null

	@BeforeAll
	fun act() {
		try {
			updater.promisePlaystatsUpdate(LibraryId(libraryId), ServiceFile(15)).toExpiringFuture().get()
		} catch (e: ExecutionException) {
			httpResponseException = e.cause as? HttpResponseException
		}
	}

	@Test
	fun thenAnHttpResponseExceptionIsThrownWithTheResponseCode() {
		assertThat(httpResponseException!!.responseCode).isEqualTo(expectedResponseCode)
	}
}
